Add basic lua implementation

SnirkImmington opened the pull request:
We have a Lua thread! We can ask it to do things! We can register Rust and Lua things for it to do!
* Dependency on `hlua`
* Lua module with thread spawned on `compositor-ready`
* Basic `LuaQuery` and `LuaResponse` enums to send messages to the thread
* `lua::send` to send messages
* Lua thread which receives requests and runs Lua code
* `lib/lua` folder with a utils library and `init.lua` to prepare the thread
* `lua::funcs` module containing Rust functions to be called from Lua (such as in `funcs::input`). This module has `#![deny(dead_code)]` so functions written here will cause a compile error if they are not registered.
* `lua::funcs::init` which adds all the Rust functions into the proper Lua APIs.
* Key binding `C-p` which prints the pointer position using Lua
* Key binding `C-l` which sleeps the Lua thread (demonstrating the main thread continues, Lua requests queue up)

This has the skeleton of the infrastructure we need for way-cooler. This is a huge part of the capability we need with the ability to have keybindings and Lua functions. After expanding on this API (add UUIDs on Lua messages, get back results, add connections to registry) we can start working on the config system, and adding more Rust and Lua functions to our API.

Timidger commented on src/lua/funcs/input.rs:
```diff
@@ -0,0 +1,31 @@
+    if in_y < 0f64 {
+        return Err("Invalid negative y parameter!".to_string());
+    }
+    //let &resolution = WlcOutput::focused().get_resolution();
```
Should probably delete this comment if it's not needed

Timidger commented on src/lua/funcs/mod.rs:
```diff
@@ -0,0 +1,33 @@
+pub fn register_libraries(lua: &mut Lua) {
+
+    // A macro for adding lua functions... could be a method but whatever
+    macro_rules! lua_function {
```
If this could be a method, it probably should be. I don't know Rust's stance on that grey line for functions that could be functions or macros, but to take a leaf out of Scheme/Lisp's book I'd say if there is something that could be a function that is instead written as a macro, you should probably write it as a function. Macros are hard to understand, and add an unnecessary level of complexity in certain cases.

Now I don't know Rust's syntax for macros, but from what I can see is that it makes a new function then automatically binds it with lua.set. Can that be a function pointer/lambda of some kind? Because it looks like you pass the function at `hlua::function0($name);`, so why do we need to have it defined? And if so, we could have it be returned as a lambda/function pointer. If you have good reasons to keep it in I'm all for hearing them, but I think this is something important to consider.

SnirkImmington commented:
Added 05100c5 in response to your comments, removed the macro for now and the comment you commented on.

Given the way we have to manually access the tables in order to register the functions, I don't think a lua macro would be as automatic, unless it took in a reference to which table we register the function with.

Timidger commented on src/lua/mod.rs:
```diff
@@ -0,0 +1,283 @@
+/// * If the lua thread is currently not running (check `lua::thread_running()`)
+/// * If the sender was unable to send, which should only happen
+/// if the lua thread isn't running.
+pub fn send(query: LuaQuery) {
```
Do we need to expose this function? `try_send` is already public, if there is a case where some rust code should panic then it can be handled by the caller (by simply unwrapping the result of `try_send`, for example).

Timidger commented on src/lua/mod.rs:
```diff
@@ -0,0 +1,283 @@
+
+        *sender = query_tx;
+        *receiver = answer_rx;
+    }
```
I'm confused, what is the purpose of this block? Aren't `sender` and `receiver` being dropped at the end of this block? They are locks on the sender/receiver, they are reassigned to `query_tx` and `answer_tx` respectively and then nothing happens?

SnirkImmington replied:
So there are four messengers:

1. Lua thread->send
2. Lua thread<- receive
3. outside->send
4. outside<-receive

The outside ones are declared as lazy statics, and the lua thread ones are moved into the thread.

> `let sender = answer_tx;`
> `let receiver = query_rx;`

However, the lazy static macro requires a thing to be initialized. One way to do it is to have them be `Option<Sender>` and `Option<Receiver>`, and ~~~unwrap~~~ `match` the values in `try_send`. I chose to be lazy and initialize them to one end of a channel (and immediately drop the other half) before the thread's `init()` fixed them.

I'll change them to `Option` types.

Timidger replied:
Ah, I see. Thanks, it would be more clear if it's an Option, though I think a comment would also have sufficed.

Timidger commented on src/lua/mod.rs:
```diff
@@ -0,0 +1,283 @@
+        *receiver = answer_rx;
+    }
+
+    thread::spawn(move || {
```
This is a pretty long block of code, should we maybe not make it a lambda? It's not particularly unreadable, just maybe it makes more sense to be a separate function? On the fence about this though, it seems fine

Timidger commented on src/lua/mod.rs:
```diff
@@ -0,0 +1,283 @@
+        lua.openlibs();
+        trace!("thread: loading way-cooler library...");
+        lua.execute_from_reader::<(), File>(
+            File::open("lib/lua/init.lua").unwrap()
```
Location of init.lua should probably be in a constant that's can easily be configured

SnirkImmington replied:
At the moment it's within the source code anyway, can configure

Timidger commented on src/lua/mod.rs:
```diff
@@ -0,0 +1,283 @@
+        // Only broadcast running after loading libs
+        *RUNNING.write().unwrap() = true;
+        trace!("thread: Testing hello world...");
+        let mut file = File::create("/tmp/init.lua").unwrap();
```
Don't assume there is a tmpfs at /tmp/. Perhaps use a crate like _tempfile_ or some other mechanism to have it write to a valid temporary place that the OS can choose.

SnirkImmington replied:
This was some testing code I wrote, should definitely be removed before the merge.

Timidger commented on src/lua/mod.rs:
```diff
@@ -0,0 +1,283 @@
+        },
+
+        LuaQuery::SetValue { name: name, val: val } => {
+        },
```
If you're not going to fill these in, might want to make it panic with a not implemented (I think there is a macro to work like that, I used it before)

Timidger commented on src/lua/mod.rs:
```diff
@@ -0,0 +1,283 @@
+        },
+
+        LuaQuery::EmptyArray(name) => {
+            
```
If you're not going to fill these in, might want to make it panic with a not implemented (I think there is a macro to work like that, I used it before)

Timidger commented on src/registry.rs:
```diff
@@ -60,3 +55,38 @@ pub fn get_or_default(name: &RegKey, value: RegVal) -> RegVal {
+/// Set a value in the registry.
+///
+/// If the key already exists, returns the old value.
+/// Returns `None` for new keys.
```
These seem like confusing return values. Why should I care what the old value is when I set it?

SnirkImmington replied:
This passes along `HashMap`'s `insert` method's return types. Most of our code will probably ignore it too.
